moshi/src/main/java/com/squareup/moshi/ClassJsonAdapter.kt

```diff
@@ -0,0 +1,229 @@
+          "undefined and unexpected behavior. Please use KotlinJsonAdapterFactory from the " +
+          "moshi-kotlin artifact or use code gen from the moshi-kotlin-codegen artifact."
+      }
+      val classFactory = ClassFactory.get<Any>(rawType)
+      val fields = sortedMapOf<String, FieldBinding<*>>()
+      var t = type
+      while (t != Any::class.java) {
+        createFieldBindings(moshi, t, fields)
+        t = Types.getGenericSuperclass(t)
+      }
+      return ClassJsonAdapter(classFactory, fields).nullSafe()
+    }
+
+    /**
+     * Throw clear error messages for the common beginner mistake of using the concrete
+     * collection classes instead of the collection interfaces, eg: ArrayList instead of List.
+     */
+    private fun Type.throwIfIsCollectionClass(collectionInterface: Class<*>) {
+      if (collectionInterface.isAssignableFrom(rawType)) {
+        throw IllegalArgumentException(
+          "No JsonAdapter for $this, you should probably use ${collectionInterface.simpleName}" +
+            " instead of ${rawType.simpleName}" +
+            " (Moshi only supports the collection interfaces by default)" +
+            " or else register a custom JsonAdapter."
+        )
+      }
+    }
+
+    /** Creates a field binding for each of declared field of `type`.  */
+    private fun createFieldBindings(
+      moshi: Moshi,
+      type: Type,
+      fieldBindings: MutableMap<String, FieldBinding<*>>
+    ) {
+      val rawType = type.rawType
+      val platformType = rawType.isPlatformType
+      for (field in rawType.declaredFields) {
+        if (!includeField(platformType, field.modifiers)) continue
+        val jsonAnnotation = field.getAnnotation(Json::class.java)
+        if (jsonAnnotation != null && jsonAnnotation.ignore) continue
+
+        // Look up a type adapter for this type.
+        val fieldType = field.genericType.resolve(type, rawType)
+        val annotations = field.jsonAnnotations
+        val fieldName = field.name
+        val adapter = moshi.adapter<Any>(
+          fieldType,
+          annotations,
+          fieldName
+        )
+
+        // Create the binding between field and JSON.
+        field.isAccessible = true
+
+        // Store it using the field's name. If there was already a field with this name, fail!
+        val jsonName = jsonAnnotation.jsonName(fieldName)
+        val fieldBinding = FieldBinding(jsonName, field, adapter)
+        val replaced = fieldBindings.put(jsonName, fieldBinding)
+        require(replaced == null) {
+          "Conflicting fields:\n" +
+            "    " +
+            requireNotNull(replaced).field +
+            "\n" +
+            "    " +
+            fieldBinding.field
+        }
+      }
+    }
+
+    /** Returns true if fields with `modifiers` are included in the emitted JSON.  */
+    private fun includeField(platformType: Boolean, modifiers: Int): Boolean {
+      if (Modifier.isStatic(modifiers) || Modifier.isTransient(modifiers)) return false
```

Comment:
Let's import all these Modifier static methods for readability

Reply:
https://github.com/square/moshi/pull/1470/commits/222c4b182ddb8b066643324d6cf6c4e253fd73bf